### PR TITLE
If order is prior to 30 days ago cancel addOrder and addItem

### PR DESCRIPTION
**Pull request by alexunder193**

### CODE REVIEW

File: includes/class-wc-skroutz-analytics-tracking.php [hunk 1]
```diff
@@ -248,4 +253,22 @@ private function calculate_tax_from_total($total, $tax_rate) {
...
+	 * @access   private
+	 */
+	private function order_is_obsolete() {
+		$thirty_days_ago = (new DateTime())->modify("-30 day");
+
+		if ( is_null($this->order->get_date_created()) ) {
+			return false;
+		}
+
+		$order_creation_date = new DateTime($this->order->get_date_created()->date('Y-m-d H:i:s'));
```
Comment (MrGoumX) [inline includes/class-wc-skroutz-analytics-tracking.php]: It seems like `date` method does not respect timezones.

```php
$this->order->get_date_created()->date(DateTimeInterface::ISO8601)
```

returns 2022-01-17T12:24:06+0000.

Instead we could use `format` which respects timezones.

```php
$this->order->get_date_created()->format(DateTimeInterface::ISO8601)
```

returns 2022-01-17T12:24:06+0200.

File: includes/class-wc-skroutz-analytics-tracking.php [hunk 2]
```diff
@@ -248,4 +253,22 @@ private function calculate_tax_from_total($total, $tax_rate) {
...
+		return ($thirty_days_ago > $order_creation_date);
```
Comment (MrGoumX) [inline includes/class-wc-skroutz-analytics-tracking.php]: Also `get_date_created` returns a WC_DateTime which inherits from DateTime. I think we can just do:

```php
return (new DateTime('- 30 days', wp_timezone())) > $this->order->get_date_created();
```

But we will have to check backwards compatibility for this.

File: includes/class-wc-skroutz-analytics-tracking.php [hunk 3]
```diff
@@ -248,4 +253,22 @@ private function calculate_tax_from_total($total, $tax_rate) {
...
+	/**
+	 * Checks if order is prior to 30 days ago
+	 *
+	 * @return boolean
+	 *
+	 * @since    1.6.9
+	 * @access   private
+	 */
+	private function order_is_obsolete() {
+		$thirty_days_ago = (new DateTime())->modify("-30 day");
```
Comment (MrGoumX) [inline includes/class-wc-skroutz-analytics-tracking.php]: And this line becomes obsolete.

File: includes/class-wc-skroutz-analytics-tracking.php [hunk 4]
```diff
@@ -248,4 +253,22 @@ private function calculate_tax_from_total($total, $tax_rate) {
 		return round( $total - $total / ( 1 + $tax_rate/100 ), 2 );
 	}
 
+	/**
+	 * Checks if order is prior to 30 days ago
+	 *
+	 * @return boolean
+	 *
+	 * @since    1.6.9
```
Comment (MrGoumX) [inline includes/class-wc-skroutz-analytics-tracking.php]: [ultra-minor] We will have to update the version here.

File: includes/class-wc-skroutz-analytics-tracking.php (hunk 1 above)
Comment (alexunder193) [reply includes/class-wc-skroutz-analytics-tracking.php]: PHP 5.4 does not accept DateTimeInterface::ISO8601. It is available in versions over 7.2. But I will keep the format method instead of date.

File: includes/class-wc-skroutz-analytics-tracking.php [hunk 5]
```diff
@@ -67,14 +67,19 @@ public function output_analytics_tracking_script() {
 		echo $analytics_script;
 	}
 
-	public function load_ecommerce_analytics( $order_id ) {
+	public function load_ecommerce_analytics($order_id ) {
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-tracking.php]: [minor] According to [wordpress php styling guide](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#space-usage) we should leave a whitespace before and after the function arguments. So I suggest to revert this change.

File: includes/class-wc-skroutz-analytics-tracking.php [hunk 6]
```diff
@@ -248,4 +253,19 @@ private function calculate_tax_from_total($total, $tax_rate) {
...
+	/**
+	 * Checks if order is prior to 30 days ago
+	 *
+	 * @return boolean
+	 *
+	 * @since    1.7.0
+	 * @access   private
+	 */
+	private function order_is_obsolete() {
+		if (is_null($this->order->get_date_created())) {
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-tracking.php]: `get_date_created` seems that is not available in our minimum supported WooCommerce version. 💥 

```
<b>Fatal error</b>
:  Call to undefined method WC_Order::get_date_created() in <b>/var/www/minimum/wp-content/plugins/skroutz-analytics-woocommerce/includes/class-wc-skroutz-analytics-tracking.php</b>
on line <b>265</b>
```

We have to find an alternative

Comment (harrisred) [reply includes/class-wc-skroutz-analytics-tracking.php]: The only alternative I found is to use the WC_Order property `order_date` that is actually the `post_date`. It seems that it does not change when the order status is updated, as it updates the property `completed_date`. However this returns a String and not a DateTime object so we have to parse it first.

File: includes/class-wc-skroutz-analytics-tracking.php [hunk 7]
```diff
@@ -248,4 +253,33 @@ private function calculate_tax_from_total($total, $tax_rate) {
...
+		else {
+			if (empty($this->order->order_date)) {
+				return false;
+			}
+
+			$order_is_obsolete = ( $thirty_days_ago > new DateTime($this->order->order_date, new DateTimeZone( '+0000' )) );
+		}
+
+		return $order_is_obsolete;
+	}
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-tracking.php]: I believe you have set `DateTimeZone( '+0000' )` as this is the default Wordpress timezone. However the Wordpress timezone can be changed from the general settings and it affects the `get_date_created`. In PHP you can actually compare DateTime objects with different timezones without an issue. So if we were supporting only the newer WooCommerce versions we wouldn't care about the timezone of the `$thirty_days_ago`, as the comparison with the order date would be correct regardless of the timezone of each DateTime object. 

The problem however lies in the older WooCommerce versions that we get the `order_date` as a string and not as a DateTime object. So we have to know the timezone setting in order to parse it correctly and then compare it with the `$thirty_days_ago`.

File: includes/class-wc-skroutz-analytics-tracking.php [hunk 8]
```diff
@@ -248,4 +253,33 @@ private function calculate_tax_from_total($total, $tax_rate) {
...
+	/**
+	 * Checks if order is prior to 30 days ago
+	 *
+	 * @return boolean
+	 *
+	 * @since    1.7.0
+	 * @access   private
+	 */
+	private function order_is_obsolete() {
+		$thirty_days_ago = new DateTime('- 30 days', new DateTimeZone( '+0000' ));
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-tracking.php]: We could set the `30` as const
```php
const OBSOLETE_ORDER_DAYS = 30;
```

File: includes/class-wc-skroutz-analytics-tracking.php (hunk 7 above)
Comment (harrisred) [reply includes/class-wc-skroutz-analytics-tracking.php]: We could refactor this like this:

```php
private function order_is_obsolete() {
  $timezone = $this->getTimezone();
  $thirty_days_ago = new DateTime('- '.self::OBSOLETE_ORDER_DAYS.' days', $timezone);

  // TODO Use only WC_Order::get_date_created when we drop support for WooCommerce < 3.0
		if ( method_exists( $this->order, 'get_date_created' ) ) {
			if ( is_null( $this->order->get_date_created() ) ) {
				return false;
			}

			$order_is_obsolete = ( $thirty_days_ago > $this->order->get_date_created() );
		}
		else {
			if ( empty( $this->order->order_date ) ) {
				return false;
			}

			$order_is_obsolete = ( $thirty_days_ago > new DateTime( $this->order->order_date, $timezone ) );
		}

		return $order_is_obsolete;
	}

	/**
	 * Get Wordpress Timezone based on settings
	 *
	 * @return DateTimeZone
	 *
	 * @since    1.7.0
	 * @access   private
	 * @see      https://wordpress.stackexchange.com/a/283094
	 */
	private function getTimezone() {
    $timezone_string = get_option( 'timezone_string' );
    if ( ! empty( $timezone_string ) ) {
        return new DateTimeZone( $timezone_string );
    }

    $offset  = get_option( 'gmt_offset' );
    $hours   = (int) $offset;
    $minutes = abs( ( $offset - (int) $offset ) * 60 );
    $offset  = sprintf( '%+03d:%02d', $hours, $minutes );

    return new DateTimeZone( $offset );
	}
```

File: includes/class-wc-skroutz-analytics-helpers.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Implements helper functions for woocommerce plugin.
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-helpers.php:8]: replace  `woocommerce` with `the`

File: includes/class-wc-skroutz-analytics-helpers.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+ */
+class WC_Skroutz_Analytics_Helpers {
+
+	/**
+	 * Get Wordpress Timezone based on settings
+	 *
+	 * @return DateTimeZone
+	 *
+	 * @since    1.7.0
+	 * @access   private
```
Comment (harrisred) [inline includes/class-wc-skroutz-analytics-helpers.php:23]: ✂️  `@access private`

**Review by harrisred — APPROVED** [review]
LGTM 👍

**Review by MrGoumX — APPROVED** [review]
+1 LGTM! 🚀
